### ISSUE TITLE
[Datetimepicker] Active bug : tag error in BVFormBtnLabelControl from BootstrapVue version 2.17.0 to 2.21.1

### ISSUE THREAD

**Issue by unrealprojet** [bootstrap-vue/bootstrap-vue#6223]
### Describe the bug

Error when using a b-form-datepicker  
TypeError: Cannot read property 'tag' of null" found in ---> <BVFormBtnLabelControl>

![image](https://user-images.githubusercontent.com/14919034/102476157-109f2980-405b-11eb-9efd-97ab5db4e924.png)

### Steps to reproduce the bug

Use bootstrap-vue 2.17.0 and after (version 2.16.0 works fine)

### Versions

**Libraries:**

- BootstrapVue: 2.17.0
- Bootstrap: 4.5.3
- Vue: 2.6.11

**Environment:**

- Device: Windows
- OS: Windows 10
- Browser: Chrome
- Version: 87

**Comment by unrealprojet** [comment]
@jackmu95 it is an ongoing issue with current version, the bug appeared in 2.17.0

**Comment by jackmu95** [comment]
@unrealprojet Can you please create a minimal reproduction of the issue on CodeSandbox or Codepen.

**Comment by Seebiscuit** [comment]
So, I drilled deep into this problem. Let me run folks through it.

First, I'll preface by saying that this issue is not limited to `BFormDatePicker`. The root of the problem starts is with `BVIconBase` (and I believe affects other components as well, since the symptom `BVIconBase` suffers from is not unique).

So starting at the `BVIconBase` source, notice that the return of `BVIconBase` render includes two children: https://github.com/bootstrap-vue/bootstrap-vue/blob/a610b445705adb36fa5f9d21fb69804b1d8c5939/src/icons/helpers/icon-base.js#L130
The implementation of `BFormDatePicker`. eventually consumes a `BVFormBtnLabelControl` which, in turn, consumes a `BIcon`. When we implement `BIcon` we eventually get to `BVIconBase`. Now back to those children in `BVIconBase`. In this render stack, starting with `BVFormBtnLabelControl`, of the two children, `$title`, `$inner`, the `$title` child is "null", because of the assignment
 here: https://github.com/bootstrap-vue/bootstrap-vue/blob/a610b445705adb36fa5f9d21fb69804b1d8c5939/src/icons/helpers/icon-base.js#L104
 Now, this is completely correct. The implementation of `BIcon` in `BVFormBtnLabelControl` does not get a `title` prop. Still, it leads to a situation where one of the children of `BVIconBase` is `null`. This is important. 
 
 So, we follow this `null` child down the stack. Eventually we end up in Vue's (2.16.2) [_createElement](https://github.com/vuejs/vue/blob/b51430f598b354ed60851bb62885539bd25de3d8/src/core/vdom/create-element.js#L47-L136) and the array of children is prepared by going through [simpleNormalizedChildren](https://github.com/vuejs/vue/blob/b51430f598b354ed60851bb62885539bd25de3d8/src/core/vdom/helpers/normalize-children.js#L18-L25):
```javascript
  if (normalizationType === ALWAYS_NORMALIZE) {
    children = normalizeChildren(children)
  } else if (normalizationType === SIMPLE_NORMALIZE) {
    children = simpleNormalizeChildren(children)   /* To simpleNormalizedChildren folks! */
  }
```
I don't want to get further in the weeds here. Suffice to say that other normalization method `normalizeChildren`, identifies the null child and removes it, returning a single item array for `children`. But, alas, our `simpleNormalizedChildren`, does not 😢. It simply returns an array, with the `$title` child the `null` reference pointer.

At this point you should be asking, why don't we go into normalizeChildren?! To understand why we earlier in the stack to [FunctionalRenderContext](https://github.com/vuejs/vue/blob/v2.6.12/src/core/vdom/create-functional-component.js#L20-L90)

and see that where `normalizationType` above gets its value is from the second line here:
```javascript
  const isCompiled = isTrue(options._compiled)
  const needNormalization = !isCompiled
```

Your next question, and our final leg before I leave you, is where is the `options._compiled` flag getting set? Seems that's what started all the trouble in the first place. The answer (which was not trivial to find) is that it's getting set by **vue-loader**. See, [normalizeComponent](https://github.com/vuejs/vue-loader/blob/v15.9.3/lib/runtime/componentNormalizer.js): 
```javascript
  // render functions
  if (render) {
    options.render = render
    options.staticRenderFns = staticRenderFns
    options._compiled = true  /* Hey SUCKER! */
  }
```
Yup, a special treatment to functional components, everywhere.

So, in short, if we're using 
 -- webpack (and who isn't) and write a
 -- SFC with a 
 -- functional component that 
 -- defines a child that can be null
 -->  Vue will explode.  

Hope you've enjoyed the show and that we can this bug _squashed_ really soon!

PS> I'd be happy to setup a repo, but sadly, I'm a complete dunce with Code Sandbox (actually just don't have the patience to put a vue cli project together). But, if anyone would point me to a Code Sandbox vue-cli boilerplate project, I'd be happy to create a repro.

**Comment by jackmu95** [comment]
@Seebiscuit I think I know how to work around it but a CodeSandbox to reproduce the issue in the first place would be very helpful.

I've created a "BootstrapVue Nuxt.js Starter Project" template on CodeSandbox:
https://codesandbox.io/s/bootstrapvue-nuxtjs-starter-project-xblbj

**Comment by jackmu95** [comment]
@Seebiscuit Can you share a minimal reproduction project with me that I can check out locally to see if the changes I have in mind indeed solve the issue?

**Comment by jackmu95** [comment]
@Seebiscuit @unrealprojet I've merged #6233 which prevents the error from happening.

But seeing the [example project](https://github.com/Seebiscuit/bootstrapvue-func-bug) to reproduce the bug I noticed that your setup from Vue.js with TypeScript is not exactly right.

When following the official [Vue.js guide](https://vuejs.org/v2/guide/typescript.html#Class-Style-Vue-Components), I have no problems using `<b-calendar>` even with v2.21.1.

**Comment by Seebiscuit** [comment]
I realized that if I add `@Component` to the App.vue class export, everything works...

```typescript
@Component  // <-- not included in repro repo
export default class App extends Vue {
...
```

Is this the standard usage you were referring to, @jackmu95?

**Comment by jackmu95** [comment]
> Is this the standard usage you were referring to, @jackmu95?

Yeah exactly.

**Comment by Seebiscuit** [comment]
So, my issue was unrelated (and self-inflicted 🤦‍♂️ ). Wondering what the underlying is issue was, then, Jack?

**Comment by Artur-Cavalcante** [comment]
> I realized that if I add `@Component` to the App.vue class export, everything works...
> 
> ```ts
> @Component  // <-- not included in repro repo
> export default class App extends Vue {
> ...
> ```
> 
> Is this the standard usage you were referring to, @jackmu95?

Work for me, thanks!
